.grit/patterns/js/move-defined-styled-components-outside-module-level.md

```diff
@@ -0,0 +1,126 @@
+---
+title: Move defined styled components outside component module level.
+---
+
+Creating a styled component inside the render method in React leads to performance issues because it dynamically generates a new component in the DOM on each render. This causes React to discard and recalculate that part of the DOM subtree every time, rather than efficiently updating only the changed parts. This can result in performance bottlenecks and unpredictable behaviour.
+
+- [reference](https://styled-components.com/docs/faqs#why-should-i-avoid-declaring-styled-components-in-the-render-method)
+
+tags: #react, #migration, #styled-component
+
+```grit
+engine marzano(0.1)
+language js
+
+or {
+    react_functional_component($props, $body) as $comp,
+    react_class_component($props, $body) as $comp
+} where {
+     $body <: contains or {
+        js"const $componentName = styled($component)`$style`",
+        js"const $componentName = styled.$tag`$style`"
+    } as $styledComponent,
+    $copy = text($styledComponent),
+    $styledComponent => .,
+    $body <: contains `return <$componentName />` => ``,
```

Comment:
This doesn't look right. Why are you wiping out the body?

Reply:
Hi @morgante ,

we will need to remove `return <$componentName />` from the `$body` because,

#### example
```
function FunctionalComponent2() {
  const Component3 = styled(FunctionalComponent)`
    color: blue;
  `
  return <Component3 /> // if the component is moved up this line is not further needed.
}
```
#### expected result
```
const Component3 = styled(FunctionalComponent)`
    color: blue;
 
function FunctionalComponent2() {

  return <Component3 /> //this needs to removed.
}

```

Reply:
I don't think that's correct. the `FunctionalComponent2` still needs a body and to return something.

Reply:
Hi @morgante , updated it now.